In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import OrdinalEncoder

In [2]:
path = 'data/healthcare-dataset-stroke-data.csv'

dataframe = pd.read_csv(path)
dataframe.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

### FP-Growth algoritm 

I'm going to use thia algorithm to find frequent patterns from the data set that can help to identify patternas that characterize the stroke. 
This algoritm work in the next way:
1. List the items of the data set
2. Scan the whole data set, count each item and remove infrequent items.
3. Group the items in order of their frequency from bigger to smaller , and create a tree where each path represents the connection between the items and common combinations.


This Algorithm is more efficient then Apriori algorithm, due to it computation cost. Apriori algorithm each time create the new group of of k items that connect and based on this groups is scan the data set one more time to create thegroups of k+1 items.(Candidate generation and than filter them by support rate).

FP-Growth Algorithm works with categorical values, and can't handle numerical ,in our data set we have the numerical values like: age, glucose level ,hypertension, bmi, stroke . 

For the items like age,glucose level and bmi i'll use the binning  and discretization techniques to convert them to categorical values.

For hypertension ,stroke and heart deciase I'll use discritization.

### PreProcessing Data

In [3]:
cp_dataframe = dataframe.copy()

In [4]:
cp_dataframe = cp_dataframe.drop(axis=1,columns="id")

For this type of algorithm, I will apply binning to the numerical values. The maximum age in the dataset is 82, though it could potentially go up to 100. Therefore, I’ve defined the age ranges as 0–29, 30–49, and 50–82.

In [5]:
bins = [0, 30, 50, 82]
labels = [ 'young adult', 'middle age', 'senior']

cp_dataframe['age_group'] = pd.cut(cp_dataframe['age'],bins = bins,labels = labels,right=True)

In [6]:
cp_dataframe

gender   age  hypertension  heart_disease ever_married      work_type  \
0       Male  67.0             0              1          Yes        Private   
1     Female  61.0             0              0          Yes  Self-employed   
2       Male  80.0             0              1          Yes        Private   
3     Female  49.0             0              0          Yes        Private   
4     Female  79.0             1              0          Yes  Self-employed   
...      ...   ...           ...            ...          ...            ...   
5105  Female  80.0             1              0          Yes        Private   
5106  Female  81.0             0              0          Yes  Self-employed   
5107  Female  35.0             0              0          Yes  Self-employed   
5108    Male  51.0             0              0          Yes        Private   
5109  Female  44.0             0              0          Yes       Govt_job   

     Residence_type  avg_glucose_level   bmi   smoking_status  stroke  \
0             Urban             228.69  36.6  formerly smoked       1   
1             Rural             202.21   NaN     never smoked       1   
2             Rural             105.92  32.5     never smoked       1   
3             Urban             171.23  34.4           smokes       1   
4             Rural             174.12  24.0     never smoked       1   
...             ...                ...   ...              ...     ...   
5105          Urban              83.75   NaN     never smoked       0   
5106          Urban             125.20  40.0     never smoked       0   
5107          Rural              82.99  30.6     never smoked       0   
5108          Rural             166.29  25.6  formerly smoked       0   
5109          Urban              85.28  26.2          Unknown       0   

       age_group  
0         senior  
1         senior  
2         senior  
3     middle age  
4         senior  
...          ...  
5105      senior  
5106      senior  
5107  middle age  
5108      senior  
5109  middle age  

[5110 rows x 12 columns]

From the dataframe we can observe couple of parameters that also have numerical values is glucose level and bmi. For them will be applied binning , for bmi is ranges of 10

In [70]:
print(max(cp_dataframe['bmi']),min(cp_dataframe['bmi']),max(cp_dataframe['avg_glucose_level']),min(cp_dataframe['avg_glucose_level']))

97.6 10.3 271.74 55.12


The maximum and minimum value for bmi is 10.3 and 97.6.We can create the range by 30.

**BMI Categories:**
  
Underweight is < 18.5  
Normal weight is 18.5 - 24.5  
Overweight is 25 - 29.9  
Obesity (Class I) is 30 - 34.9  
Obesity (Class II) is 35 - 39.9  
Obesity (class III) is 40+  

**Average Glucose Levels:**

Adults and Childern 13+ normal is 70-140 mg/dl  
Children under 13 normal is 70-180 mg/dl


In [7]:
bmi_mean_by_age = cp_dataframe.groupby('age_group')['bmi'].mean().to_dict()

/var/folders/28/pmh07v693dn2c3l0jdz5zvlw0000gn/T/ipykernel_5120/754360870.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  bmi_mean_by_age = cp_dataframe.groupby('age_group')['bmi'].mean().to_dict()


Create the bins for glucose level and bmi. For glucose levele i have bins of the size 20, and for bmi 10.

In [8]:
for key,val in bmi_mean_by_age.items():
    mask = (cp_dataframe['age_group'] == key) & (cp_dataframe['bmi'].isna())
    cp_dataframe.loc[mask,'bmi'] = val
    
cp_dataframe['bmi'] = cp_dataframe['bmi'].round(2)

#discretization and smoothing for bmi
# Define bin edges from 10 to 100 with a step of 10
bin_edges_bmi = np.arange(10, 100 + 10, 10)  # [10, 20, 30, ..., 100]

bin_edges_glucose = np.arange(55,275+90,90)
# Create BMI bins
cp_dataframe['bmi_bin'] = pd.cut(cp_dataframe['bmi'], bins=bin_edges_bmi, include_lowest=True)
#Create Glucose Bins 
cp_dataframe['avg_glucose_bin']= pd.cut(cp_dataframe['avg_glucose_level'], bins=bin_edges_glucose, include_lowest=True)




In [9]:
cp_dataframe

gender   age  hypertension  heart_disease ever_married      work_type  \
0       Male  67.0             0              1          Yes        Private   
1     Female  61.0             0              0          Yes  Self-employed   
2       Male  80.0             0              1          Yes        Private   
3     Female  49.0             0              0          Yes        Private   
4     Female  79.0             1              0          Yes  Self-employed   
...      ...   ...           ...            ...          ...            ...   
5105  Female  80.0             1              0          Yes        Private   
5106  Female  81.0             0              0          Yes  Self-employed   
5107  Female  35.0             0              0          Yes  Self-employed   
5108    Male  51.0             0              0          Yes        Private   
5109  Female  44.0             0              0          Yes       Govt_job   

     Residence_type  avg_glucose_level    bmi   smoking_status  stroke  \
0             Urban             228.69  36.60  formerly smoked       1   
1             Rural             202.21  30.67     never smoked       1   
2             Rural             105.92  32.50     never smoked       1   
3             Urban             171.23  34.40           smokes       1   
4             Rural             174.12  24.00     never smoked       1   
...             ...                ...    ...              ...     ...   
5105          Urban              83.75  30.67     never smoked       0   
5106          Urban             125.20  40.00     never smoked       0   
5107          Rural              82.99  30.60     never smoked       0   
5108          Rural             166.29  25.60  formerly smoked       0   
5109          Urban              85.28  26.20          Unknown       0   

       age_group       bmi_bin  avg_glucose_bin  
0         senior  (30.0, 40.0]   (145.0, 235.0]  
1         senior  (30.0, 40.0]   (145.0, 235.0]  
2         senior  (30.0, 40.0]  (54.999, 145.0]  
3     middle age  (30.0, 40.0]   (145.0, 235.0]  
4         senior  (20.0, 30.0]   (145.0, 235.0]  
...          ...           ...              ...  
5105      senior  (30.0, 40.0]  (54.999, 145.0]  
5106      senior  (30.0, 40.0]  (54.999, 145.0]  
5107  middle age  (30.0, 40.0]  (54.999, 145.0]  
5108      senior  (20.0, 30.0]   (145.0, 235.0]  
5109  middle age  (20.0, 30.0]  (54.999, 145.0]  

[5110 rows x 14 columns]

In [10]:
cp_dataframe = cp_dataframe.drop(axis =1, columns=['avg_glucose_level','age','bmi'])

In [11]:
categorical_columns = cp_dataframe.select_dtypes(include=['object', 'category']).columns.tolist()
categorical_columns

['gender',
 'ever_married',
 'work_type',
 'Residence_type',
 'smoking_status',
 'age_group',
 'bmi_bin',
 'avg_glucose_bin']

In [121]:
encoder =OrdinalEncoder()

dataframe_cp = cp_dataframe.copy()
dataframe_cp[categorical_columns] = encoder.fit_transform(cp_dataframe[categorical_columns])

In [122]:
data_no_stroke = dataframe_cp.drop(columns='stroke')
stroke_column  = dataframe_cp['stroke']

smote = SMOTE(random_state=42)
X_resampled, y_resampled = smote.fit_resample(data_no_stroke, stroke_column)

In [123]:
X_resampled[categorical_columns] = encoder.inverse_transform(X_resampled[categorical_columns])

In [124]:
data = X_resampled.join(y_resampled,how='inner')
data['bmi_bin'] = data['bmi_bin'].astype('category')
data['avg_glucose_bin'] = data['avg_glucose_bin'].astype('category')

In [125]:
#lets one-hot encode the data set 
df_encoded = pd.get_dummies(data)

# Step 3: Convert to True/False format (optional, but cleaner for FP-Growth)
df_encoded = df_encoded.astype(bool)

In [126]:
df_encoded

hypertension  heart_disease  stroke  gender_Female  gender_Male  \
0            False           True    True          False         True   
1            False          False    True           True        False   
2            False           True    True          False         True   
3            False          False    True           True        False   
4             True          False    True           True        False   
...            ...            ...     ...            ...          ...   
9717          True          False    True           True        False   
9718         False          False    True           True        False   
9719         False          False    True          False         True   
9720         False          False    True           True        False   
9721         False          False    True           True        False   

      gender_Other  ever_married_No  ever_married_Yes  work_type_Govt_job  \
0            False            False              True               False   
1            False            False              True               False   
2            False            False              True               False   
3            False            False              True               False   
4            False            False              True               False   
...            ...              ...               ...                 ...   
9717         False            False              True               False   
9718         False            False              True                True   
9719         False            False              True               False   
9720         False            False              True               False   
9721         False            False              True               False   

      work_type_Never_worked  ...  bmi_bin_(20.0, 30.0]  bmi_bin_(30.0, 40.0]  \
0                      False  ...                 False                  True   
1                      False  ...                 False                  True   
2                      False  ...                 False                  True   
3                      False  ...                 False                  True   
4                      False  ...                  True                 False   
...                      ...  ...                   ...                   ...   
9717                   False  ...                  True                 False   
9718                   False  ...                  True                 False   
9719                   False  ...                 False                  True   
9720                   False  ...                  True                 False   
9721                   False  ...                 False                  True   

      bmi_bin_(40.0, 50.0]  bmi_bin_(50.0, 60.0]  bmi_bin_(60.0, 70.0]  \
0                    False                 False                 False   
1                    False                 False                 False   
2                    False                 False                 False   
3                    False                 False                 False   
4                    False                 False                 False   
...                    ...                   ...                   ...   
9717                 False                 False                 False   
9718                 False                 False                 False   
9719                 False                 False                 False   
9720                 False                 False                 False   
9721                 False                 False                 False   

      bmi_bin_(70.0, 80.0]  bmi_bin_(90.0, 100.0]  \
0                    False                  False   
1                    False                  False   
2                    False                  False   
3                    False                  False   
4                    False                  False   
...                    .

In [14]:
#lets use the FP-Growth algorithm to check frequent patterns can we observe from the data and them retreive the association rules 
from mlxtend.frequent_patterns import fpgrowth, association_rules

In [15]:
frequent_itemsets = fpgrowth(df_encoded, min_support=0.4, use_colnames=True,verbose=1)


NameError: name 'df_encoded' is not defined

In [ ]:
frequent_itemsets 

support                                           itemsets
0   0.757560                                 (ever_married_Yes)
1   0.649969                                 (age_group_senior)
2   0.604814                                (work_type_Private)
3   0.500000                                           (stroke)
4   0.487760                             (Residence_type_Urban)
5   0.618186                                    (gender_Female)
6   0.512240                             (Residence_type_Rural)
7   0.790064                  (avg_glucose_bin_(54.999, 145.0])
8   0.509977                             (bmi_bin_(20.0, 30.0])
9   0.569327  (ever_married_Yes, avg_glucose_bin_(54.999, 14...
10  0.578070               (ever_married_Yes, age_group_senior)
11  0.468422  (age_group_senior, avg_glucose_bin_(54.999, 14...
12  0.409072  (ever_married_Yes, age_group_senior, avg_gluco...
13  0.487657              (work_type_Private, ever_married_Yes)
14  0.456593  (work_type_Private, avg_glucose_bin_(54.999, 1...
15  0.454433                         (stroke, age_group_senior)
16  0.435301                         (stroke, ever_married_Yes)
17  0.468011                  (ever_married_Yes, gender_Female)
18  0.402386                  (age_group_senior, gender_Female)
19  0.501954   (gender_Female, avg_glucose_bin_(54.999, 145.0])
20  0.402283  (avg_glucose_bin_(54.999, 145.0], Residence_ty...
21  0.439004  (bmi_bin_(20.0, 30.0], avg_glucose_bin_(54.999...

In [129]:
frequent_itemsets['itemsets']

0                                    (ever_married_Yes)
1                                    (age_group_senior)
2                                   (work_type_Private)
3                                              (stroke)
4                                (Residence_type_Urban)
5                                       (gender_Female)
6                                (Residence_type_Rural)
7                     (avg_glucose_bin_(54.999, 145.0])
8                                (bmi_bin_(20.0, 30.0])
9     (ever_married_Yes, avg_glucose_bin_(54.999, 14...
10                 (ever_married_Yes, age_group_senior)
11    (age_group_senior, avg_glucose_bin_(54.999, 14...
12    (ever_married_Yes, age_group_senior, avg_gluco...
13                (work_type_Private, ever_married_Yes)
14    (work_type_Private, avg_glucose_bin_(54.999, 1...
15                           (stroke, age_group_senior)
16                           (stroke, ever_married_Yes)
17                    (ever_married_Yes, gender_

In [130]:
# Step 5: Generate rules
rules = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.6)

In [131]:
rules

antecedents  \
0                                  (ever_married_Yes)   
1                   (avg_glucose_bin_(54.999, 145.0])   
2                                  (ever_married_Yes)   
3                                  (age_group_senior)   
4                                  (age_group_senior)   
5                (ever_married_Yes, age_group_senior)   
6   (ever_married_Yes, avg_glucose_bin_(54.999, 14...   
7   (age_group_senior, avg_glucose_bin_(54.999, 14...   
8                                  (age_group_senior)   
9                                 (work_type_Private)   
10                                 (ever_married_Yes)   
11                                (work_type_Private)   
12                                           (stroke)   
13                                 (age_group_senior)   
14                                           (stroke)   
15                                 (ever_married_Yes)   
16                                    (gender_Female)   
17                                 (age_group_senior)   
18                                    (gender_Female)   
19                                    (gender_Female)   
20                  (avg_glucose_bin_(54.999, 145.0])   
21                             (Residence_type_Rural)   
22                             (bmi_bin_(20.0, 30.0])   

                                          consequents  antecedent support  \
0                   (avg_glucose_bin_(54.999, 145.0])            0.757560   
1                                  (ever_married_Yes)            0.790064   
2                                  (age_group_senior)            0.757560   
3                                  (ever_married_Yes)            0.649969   
4                   (avg_glucose_bin_(54.999, 145.0])            0.649969   
5                   (avg_glucose_bin_(54.999, 145.0])            0.578070   
6                                  (age_group_senior)            0.569327   
7                                  (ever_married_Yes)            0.468422   
8   (ever_married_Yes, avg_glucose_bin_(54.999, 14...            0.649969   
9                                  (ever_married_Yes)            0.604814   
10                                (work_type_Private)            0.757560   
11                  (avg_glucose_bin_(54.999, 145.0])            0.604814   
12                                 (age_group_senior)            0.500000   
13                                           (stroke)            0.649969   
14                                 (ever_married_Yes)            0.500000   
15                                    (gender_Female)            0.757560   
16                                 (ever_married_Yes)            0.618186   
17                                    (gender_Female)            0.649969   
18                                 (age_group_senior)            0.618186   
19                  (avg_glucose_bin_(54.999, 145.0])            0.618186   
20                                    (gender_Female)            0.790064   
21                  (avg_glucose_bin_(54.999, 145.0])            0.512240   
22                  (avg_glucose_bin_(54.999, 145.0])            0.509977   

    consequent support   support  confidence      lift  representativity  \
0             0.790064  0.569327    0.751527  0.951224               1.0   
1             0.757560  0.569327    0.720609  0.951224               1.0   
2             0.649969  0.578070    0.763069  1.174007               1.0   
3             0.757560  0.578070    0.889381  1.174007               1.0   
4             0.790064  0.468422    0.720684  0.912184               1.0   
5             0.790064  0.409072    0.707651  0.895689               1.0   
6             0.649969  0.409072    0.718519  1.105466               1.0   
7             0.757560  0.409072    0.873298  1.152777               1.0   
8             0.569327  0.409072    0.629372  1.105466               1.0   
9             0.757560  0.487657    0.806293  1.064328          

I initially used the SMOTE technique to generate more data for the minority class (stroke cases) and then used this augmented dataset to create association rules. However, there were only a few association rules specifically related to stroke. Now, I want to take a different approach by using only the original stroke-positive data — before applying SMOTE — to generate association rules and compare the results.

In [12]:
stroke_positive_data = cp_dataframe[cp_dataframe["stroke"]==1]

In [13]:
stroke_positive_data

gender  hypertension  heart_disease ever_married      work_type  \
0      Male             0              1          Yes        Private   
1    Female             0              0          Yes  Self-employed   
2      Male             0              1          Yes        Private   
3    Female             0              0          Yes        Private   
4    Female             1              0          Yes  Self-employed   
..      ...           ...            ...          ...            ...   
244    Male             0              0          Yes        Private   
245  Female             0              0           No       children   
246  Female             0              0          Yes  Self-employed   
247    Male             1              0          Yes  Self-employed   
248  Female             0              0          Yes        Private   

    Residence_type   smoking_status  stroke    age_group        bmi_bin  \
0            Urban  formerly smoked       1       senior   (30.0, 40.0]   
1            Rural     never smoked       1       senior   (30.0, 40.0]   
2            Rural     never smoked       1       senior   (30.0, 40.0]   
3            Urban           smokes       1   middle age   (30.0, 40.0]   
4            Rural     never smoked       1       senior   (20.0, 30.0]   
..             ...              ...     ...          ...            ...   
244          Rural          Unknown       1       senior   (30.0, 40.0]   
245          Rural          Unknown       1  young adult   (30.0, 40.0]   
246          Rural  formerly smoked       1       senior   (20.0, 30.0]   
247          Rural          Unknown       1       senior   (30.0, 40.0]   
248          Rural          Unknown       1       senior  (9.999, 20.0]   

     avg_glucose_bin  
0     (145.0, 235.0]  
1     (145.0, 235.0]  
2    (54.999, 145.0]  
3     (145.0, 235.0]  
4     (145.0, 235.0]  
..               ...  
244  (54.999, 145.0]  
245  (54.999, 145.0]  
246  (54.999, 145.0]  
247  (54.999, 145.0]  
248  (54.999, 145.0]  

[249 rows x 11 columns]

We have got only 249 samples, lets check if there any rules could be retrieved .

In [16]:
#lets one-hot encode the data set 
stroke_encoded = pd.get_dummies(stroke_positive_data)

# Step 3: Convert to True/False format (optional, but cleaner for FP-Growth)
stroke_encoded = stroke_encoded.astype(bool)

In [17]:
frequent_itemsets_stroke = fpgrowth(stroke_encoded, min_support=0.4, use_colnames=True,verbose=1)


11 itemset(s) from tree conditioned on items ()
0 itemset(s) from tree conditioned on items (stroke)
1 itemset(s) from tree conditioned on items (age_group_senior)
3 itemset(s) from tree conditioned on items (ever_married_Yes)
3 itemset(s) from tree conditioned on items (work_type_Private)
0 itemset(s) from tree conditioned on items (work_type_Private, stroke)
1 itemset(s) from tree conditioned on items (work_type_Private, ever_married_Yes)
3 itemset(s) from tree conditioned on items (work_type_Private, age_group_senior)
3 itemset(s) from tree conditioned on items (Residence_type_Urban)
0 itemset(s) from tree conditioned on items (Residence_type_Urban, stroke)
1 itemset(s) from tree conditioned on items (Residence_type_Urban, age_group_senior)
3 itemset(s) from tree conditioned on items (Residence_type_Urban, ever_married_Yes)
3 itemset(s) from tree conditioned on items (bmi_bin_(30.0, 40.0])
0 itemset(s) from tree conditioned on items (bmi_bin_(30.0, 40.0], stroke)
1 itemset(s) from t

In [20]:
frequent_itemsets_stroke['itemsets']

0                                              (stroke)
1                                    (age_group_senior)
2                                    (ever_married_Yes)
3                                   (work_type_Private)
4                                (Residence_type_Urban)
                            ...                        
56    (avg_glucose_bin_(54.999, 145.0], age_group_se...
57    (avg_glucose_bin_(54.999, 145.0], age_group_se...
58                       (bmi_bin_(20.0, 30.0], stroke)
59             (age_group_senior, bmi_bin_(20.0, 30.0])
60     (age_group_senior, bmi_bin_(20.0, 30.0], stroke)
Name: itemsets, Length: 61, dtype: object

In [18]:
# Step 5: Generate rules
rules_stroe = association_rules(frequent_itemsets_stroke, metric='confidence', min_threshold=0.6)

/Users/alexakh/.pyenv/versions/3.11.0/lib/python3.11/site-packages/mlxtend/frequent_patterns/association_rules.py:186: RuntimeWarning: invalid value encountered in divide
  cert_metric = np.where(certainty_denom == 0, 0, certainty_num / certainty_denom)


In [19]:
rules_stroe

antecedents                 consequents  \
0                          (age_group_senior)                    (stroke)   
1                                    (stroke)          (age_group_senior)   
2                          (ever_married_Yes)                    (stroke)   
3                                    (stroke)          (ever_married_Yes)   
4                          (age_group_senior)          (ever_married_Yes)   
..                                        ...                         ...   
125                    (bmi_bin_(20.0, 30.0])                    (stroke)   
126                    (bmi_bin_(20.0, 30.0])          (age_group_senior)   
127  (age_group_senior, bmi_bin_(20.0, 30.0])                    (stroke)   
128            (bmi_bin_(20.0, 30.0], stroke)          (age_group_senior)   
129                    (bmi_bin_(20.0, 30.0])  (age_group_senior, stroke)   

     antecedent support  consequent support   support  confidence      lift  \
0              0.907631            1.000000  0.907631    1.000000  1.000000   
1              1.000000            0.907631  0.907631    0.907631  1.000000   
2              0.883534            1.000000  0.883534    1.000000  1.000000   
3              1.000000            0.883534  0.883534    0.883534  1.000000   
4              0.907631            0.883534  0.811245    0.893805  1.011625   
..                  ...                 ...       ...         ...       ...   
125            0.445783            1.000000  0.445783    1.000000  1.000000   
126            0.445783            0.907631  0.405622    0.909910  1.002511   
127            0.405622            1.000000  0.405622    1.000000  1.000000   
128            0.445783            0.907631  0.405622    0.909910  1.002511   
129            0.445783            0.907631  0.405622    0.909910  1.002511   

     representativity  leverage  conviction  zhangs_metric   jaccard  \
0                 1.0  0.000000         inf       0.000000  0.907631   
1                 1.0  0.000000    1.000000       0.000000  0.907631   
2                 1.0  0.000000         inf       0.000000  0.883534   
3                 1.0  0.000000    1.000000       0.000000  0.883534   
4                 1.0  0.009322    1.096720       0.124408  0.827869   
..                ...       ...         ...            ...       ...   
125               1.0  0.000000         inf       0.000000  0.445783   
126               1.0  0.001016    1.025301       0.004520  0.427966   
127               1.0  0.000000         inf       0.000000  0.405622   
128               1.0  0.001016    1.025301       0.004520  0.427966   
129               1.0  0.001016    1.025301       0.004520  0.427966   

     certainty  kulczynski  
0     0.000000    0.953815  
1     0.000000    0.953815  
2     0.000000    0.941767  
3     0.000000    0.941767  
4     0.088190    0.905994  
..         ...         ...  
125   0.000000    0.722892  
126   0.024677    0.678406  
127   0.000000    0.702811  
128   0.024677    0.678406  
129   0.024677    0.678406  

[130 rows x 14 columns]

In [25]:
rules_stroe[rules_stroe['consequents']==frozenset({'stroke'})]

antecedents consequents  \
0                                   (age_group_senior)    (stroke)   
2                                   (ever_married_Yes)    (stroke)   
6                 (age_group_senior, ever_married_Yes)    (stroke)   
12                                 (work_type_Private)    (stroke)   
16               (work_type_Private, ever_married_Yes)    (stroke)   
22               (work_type_Private, age_group_senior)    (stroke)   
28   (work_type_Private, ever_married_Yes, age_grou...    (stroke)   
34                              (Residence_type_Urban)    (stroke)   
37            (age_group_senior, Residence_type_Urban)    (stroke)   
40            (ever_married_Yes, Residence_type_Urban)    (stroke)   
46   (age_group_senior, ever_married_Yes, Residence...    (stroke)   
53                              (bmi_bin_(30.0, 40.0])    (stroke)   
56            (bmi_bin_(30.0, 40.0], age_group_senior)    (stroke)   
59            (bmi_bin_(30.0, 40.0], ever_married_Yes)    (stroke)   
62                                       (gender_Male)    (stroke)   
65                     (age_group_senior, gender_Male)    (stroke)   
68                     (ever_married_Yes, gender_Male)    (stroke)   
71                                     (gender_Female)    (stroke)   
74                   (gender_Female, age_group_senior)    (stroke)   
77                   (gender_Female, ever_married_Yes)    (stroke)   
83   (gender_Female, age_group_senior, ever_married...    (stroke)   
90                              (Residence_type_Rural)    (stroke)   
94            (Residence_type_Rural, ever_married_Yes)    (stroke)   
96            (age_group_senior, Residence_type_Rural)    (stroke)   
99                   (avg_glucose_bin_(54.999, 145.0])    (stroke)   
105  (avg_glucose_bin_(54.999, 145.0], age_group_se...    (stroke)   
110  (avg_glucose_bin_(54.999, 145.0], ever_married...    (stroke)   
118  (avg_glucose_bin_(54.999, 145.0], age_group_se...    (stroke)   
125                             (bmi_bin_(20.0, 30.0])    (stroke)   
127           (age_group_senior, bmi_bin_(20.0, 30.0])    (stroke)   

     antecedent support  consequent support   support  confidence  lift  \
0              0.907631                 1.0  0.907631         1.0   1.0   
2              0.883534                 1.0  0.883534         1.0   1.0   
6              0.811245                 1.0  0.811245         1.0   1.0   
12             0.598394                 1.0  0.598394         1.0   1.0   
16             0.542169                 1.0  0.542169         1.0   1.0   
22             0.534137                 1.0  0.534137         1.0   1.0   
28             0.481928                 1.0  0.481928         1.0   1.0   
34             0.542169                 1.0  0.542169         1.0   1.0   
37             0.497992                 1.0  0.497992         1.0   1.0   
40             0.457831                 1.0  0.457831         1.0   1.0   
46             0.425703                 1.0  0.425703         1.0   1.0   
53             0.469880                 1.0  0.469880         1.0   1.0   
56             0.429719                 1.0  0.429719         1.0   1.0   
59             0.409639                 1.0  0.409639         1.0   1.0   
62             0.433735                 1.0  0.433735         1.0   1.0   
65             0.409639                 1.0  0.409639         1.0   1.0   
68             0.401606                 1.0  0.401606         1.0   1.0   
71             0.566265                 1.0  0.566265         1.0   1.0   
74             0.497992                 1.0  0.497992         1.0   1.0   
77             0.481928                 1.0  0.481928         1.0   1.0   
83             0.425703                 1.0  0.425703         1.0   1.0   
90             0.457831                 1.0  0.457831         1.0   1.0   
94             0.425703                 1.0  0.425703         1.0   1.0   
96             0.409639                 1.0  0.409639         1.0   1.0 